### ISSUE TITLE
[BUG]CryptographyClient failed to gettoken when SNI configured but it works if i upload the certificate to AAD app

### ISSUE THREAD

**Issue by PPrabinKumarPatro** [Azure/azure-sdk-for-net#16298]
**Describe the bug**
CryptographyClient failed to gettoken when SNI configured but it works if i upload the certificate to AAD app
This is while doing client side encryption for storage blob data using keyvault key

**Expected behavior**
When subject name Issuer is configured, client side encryption should work

**Actual behavior (include Exception or Stack Trace)**
What is the actual behavior?

**To Reproduce**
Steps to reproduce the behavior (include a code snippet, screenshot, or any additional information that might help us reproduce the issue)
 Below is the details
 
If i upload cert to AAD app instead using SNI,  it works  but we need with SNI, below is the token credential
 
 ```csharp
TokenCredentialOptions tokenCredentialOptions = new TokenCredentialOptions();
                tokenCredentialOptions.AuthorityHost = new Uri("https://login.windows.net/");
                X509Certificate2 certificate = CertificateHelper.GetCertificateBySubject(out var message, config.GetRequiredSetting("AAD_Certificate_SubjectName"));
                TokenCredential credential = new ClientCertificateCredential(
                    "72f988bf-86f1-41af-91ab-2d7cd011db47",
                    "4fc6e86a-8a62-4ba8-9254-a20db9b6c6f9",
                    certificate,
                    tokenCredentialOptions);
 
//Create CryptographyClient using Key Vault Key
                CryptographyClient cryptographyClient = new CryptographyClient(clientSideKeyVaultKeyUri, credential);
``` 
 
SO i thought issue in generating a token while SNI configured and SNI was not configured properly, so i tried to generate a token using below code and its successful, that mean SNI was configured properly
 
```csharp
var _clientAssertionCertificate = new ClientAssertionCertificate("4fc6e86a-8a62-4ba8-9254-a20db9b6c6f9", certificate);

  var authContext = new AuthenticationContext(new Uri( "https://login.windows.net/",
                   "72f988bf-86f1-41af-91ab-2d7cd011db47");

            AuthenticationResult token = await authContext.AcquireTokenAsync(  "https://vault.azure.net", _clientAssertionCertificate, true);
``` 
 
So not able to figure out what is the issue

[5:47 PM] Pitta Prabin Patro
    Also i checked SNI was configured properly
-----------------------------------------------------
Deserialized JSON Data for TrustedCertificateSubjects
-----------------------------------------------------
[
  {​​​​​​​
    "authorityId": "00000000-0000-0000-0000-000000000001",
    "subjectName": "billingpipeline-int-aad.int.applicationinsights.ame"
  }​​​​​​​
]

    So suspecting bug how CryptographyClient is generating token or some parameter is missing

    which is not documented or i am not able to find that

    or i am missing something

**CALLSTACK**
call stack is 
   at Azure.Identity.CredentialDiagnosticScope.FailWrapAndThrow(Exception ex)
   at Azure.Identity.ClientCertificateCredential.GetToken(TokenRequestContext requestContext, CancellationToken cancellationToken)
   at Azure.Security.KeyVault.ChallengeBasedAuthenticationPolicy.<AuthenticateRequestAsync>d__9.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Azure.Security.KeyVault.ChallengeBasedAuthenticationPolicy.<ProcessCoreAsync>d__8.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ValueTaskAwaiter.GetResult()
   at Azure.Core.Pipeline.TaskExtensions.EnsureCompleted(ValueTask task)
   at Azure.Security.KeyVault.ChallengeBasedAuthenticationPolicy.Process(HttpMessage message, ReadOnlyMemory`1 pipeline)
   at Azure.Core.Pipeline.HttpPipelinePolicy.ProcessNext(HttpMessage message, ReadOnlyMemory`1 pipeline)
   at Azure.Core.Pipeline.RetryPolicy.<ProcessAsync>d__11.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Azure.Core.Pipeline.RetryPolicy.<ProcessAsync>d__11.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ValueTaskAwaiter.GetResult()
   at Azure.Core.Pipeline.TaskExtensions.EnsureCompleted(ValueTask task)
   at Azure.Core.Pipeline.RetryPolicy.Process(HttpMessage message, ReadOnlyMemory`1 pipeline)
   at Azure.Core.Pipeline.HttpPipelinePolicy.ProcessNext(HttpMessage message, ReadOnlyMemory`1 pipeline)
   at Azure.Core.Pipeline.HttpPipelineSynchronousPolicy.Process(HttpMessage message, ReadOnlyMemory`1 pipeline)
   at Azure.Core.Pipeline.HttpPipelinePolicy.ProcessNext(HttpMessage message, ReadOnlyMemory`1 pipeline)
   at Azure.Core.Pipeline.HttpPipelineSynchronousPolicy.Process(HttpMessage message, ReadOnlyMemory`1 pipeline)
   at Azure.Core.Pipeline.HttpPipelinePolicy.ProcessNext(HttpMessage message, ReadOnlyMemory`1 pipeline)
   at Azure.Core.Pipeline.HttpPipelineSynchronousPolicy.Process(HttpMessage message, ReadOnlyMemory`1 pipeline)
   at Azure.Core.Pipeline.HttpPipeline.Send(HttpMessage message, CancellationToken cancellationToken)
   at Azure.Core.Pipeline.HttpPipeline.SendRequest(Request request, CancellationToken cancellationToken)
   at Azure.Security.KeyVault.KeyVaultPipeline.SendRequest(Request request, CancellationToken cancellationToken)
   at Azure.Security.KeyVault.KeyVaultPipeline.SendRequest[TResult](RequestMethod method, Func`1 resultFactory, CancellationToken cancellationToken, String[] path)
   at Azure.Security.KeyVault.Keys.Cryptography.RemoteCryptographyClient.GetKey(CancellationToken cancellationToken)
   at Azure.Security.KeyVault.Keys.Cryptography.CryptographyClient.Initialize(String operation, CancellationToken cancellationToken)
   at Azure.Security.KeyVault.Keys.Cryptography.CryptographyClient.WrapKey(KeyWrapAlgorithm algorithm, Byte[] key, CancellationToken cancellationToken)
   at Azure.Security.KeyVault.Keys.Cryptography.CryptographyClient.Azure.Core.Cryptography.IKeyEncryptionKey.WrapKey(String algorithm, ReadOnlyMemory`1 key, CancellationToken cancellationToken)
   at Azure.Storage.Cryptography.Models.EncryptionData.<CreateInternalV1_0>d__20.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
   at Azure.Storage.Cryptography.ClientSideEncryptor.<EncryptInternal>d__3.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
   at Azure.Storage.Blobs.BlobClientSideEncryptor.<ClientSideEncryptInternal>d__2.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
   at Azure.Storage.Blobs.BlobClient.<StagedUploadInternal>d__29.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at Azure.Core.Pipeline.TaskExtensions.EnsureCompleted[T](Task`1 task)
   at Azure.Storage.Blobs.BlobClient.Upload(Stream content, BlobHttpHeaders httpHeaders, IDictionary`2 metadata, BlobRequestConditions conditions, IProgress`1 progressHandler, Nullable`1 accessTier, StorageTransferOptions transferOptions, CancellationToken cancellationToken)
   at Azure.Storage.Blobs.BlobClient.Upload(Stream content, Boolean overwrite, CancellationToken cancellationToken)
   at Azure.Storage.Blobs.BlobClient.Upload(Stream content, CancellationToken cancellationToken)
   at Azure.Storage.Blobs.BlobClient.Upload(Stream content)
   at Microsoft.AzureMonitor.Billing.LogAnalyticsCollectorBusiness.LogAnalyticsCollectorManager.<StoreEventHubInfoToBlobAsync2>d__16.MoveNext() in C:\repos\Mgmt-AzureMonitor-BillingPipeline\src\LogAnalyticsCollectorService\Business\LogAnalyticsCollectorManager.cs:line 376

**FULL CODE SNIPPET:**
1. 
                TokenCredentialOptions tokenCredentialOptions = new TokenCredentialOptions();
                tokenCredentialOptions.AuthorityHost = new Uri("https://login.windows.net/");
                X509Certificate2 certificate = CertificateHelper.GetCertificateBySubject(out var message, config.GetRequiredSetting("AAD_Certificate_SubjectName"));
                TokenCredential credential = new ClientCertificateCredential(
                    "72f988bf-86f1-41af-91ab-2d7cd011db47",
                    "4fc6e86a-8a62-4ba8-9254-a20db9b6c6f9",
                    certificate,
                    tokenCredentialOptions);
2.  //Get Uri for Key Vault key
                Uri clientSideKeyVaultKeyUri = new Uri("https://labilling-int-eus.vault.azure.net/keys/TestPrabin");

                //Create CryptographyClient using Key Vault Key
                CryptographyClient cryptographyClient = new CryptographyClient(clientSideKeyVaultKeyUri, credential);

                //Set up Client Side Encryption Options used for Client Side Encryption
                ClientSideEncryptionOptions clientSideOptions = new ClientSideEncryptionOptions(ClientSideEncryptionVersion.V1_0)
                {
                    // KeyEncryptionKey = cryptographyClient,
                    //KeyWrapAlgorithm = "RS256"
                    KeyEncryptionKey = cryptographyClient,
                   // KeyResolver = keyResolver,
                    // string the storage client will use when calling IKeyEncryptionKey.WrapKey()
                    KeyWrapAlgorithm = "RSA1_5"
                };

                //Create Blob Service Client
                BlobServiceClient blobServiceClient = new BlobServiceClient(storageConnectionString);

                BlobContainerClient containerClient = blobServiceClient.GetBlobContainerClient("example");
                //Create BlobClient with Client Side Encryption Options to upload client side encrypted data
                BlobClient blobClient = containerClient.GetBlobClient("test").WithClientSideEncryptionOptions(clientSideOptions);
                blobClient.Upload(stream);

**Environment:**
 - Name and version of the Library package used: [e.g. Azure.Storage.Blobs 12.2.0] 
Azure.Storage.Blobs : 12.6.0
Azure.Identity is 1.2.3
Azure.Security.KeyVault.Keys 4.1.0

 - Hosting platform or OS and .NET runtime version (`dotnet --info` output for .NET Core projects): [e.g. Azure AppService or Windows 10 .NET Framework 4.8]
.Net core
 - IDE and version : [e.g. Visual Studio 16.3]
Visual studio 2019

**Comment by jsquire** [comment]
Thank you for your feedback.  Tagging and routing to the team member best able to assist.

**Comment by jsquire** [comment]
@heaths:  I'm not sure if this falls into the purview of KeyVault or Identity.  Do you mind taking a look and re-routing if needed?

//fyi: @schaabs

**Comment by schaabs** [comment]
@PPrabinKumarPatro Thanks for filing this detailed issue. We recently added support for Subject Name + Issuer authentication to the `ClientCertificateCredential`, however, this feature is still only available in beta, Azure.Identity 1.2.3 doesn't support it.  If you would like to try out this feature in our beta release it is available in version 1.3.0-beta.2. Once you've upgraded to this version you will also need to update your code to enable SN+I authentication, by setting the `IncludeX5CCliamHeader` property to true on the `ClientCertificateCredentialOptions`.

```c# 
var credential = new ClientCertificateCredential(tenantId, clientId, certPath, new ClientCertificateCredentialOptions { IncludeX5CCliamHeader = true });
```

I'm closing this issue but if you still are having trouble getting this to work with the beta package please feel free to reopen it or open anothoer. The feature is still in beta so any feedback you have would be greatly appreciated. Just a heads up we will be releasing Azure.Identity 1.3.0 next month. In that release we intend to renamed the `IncludeX5CCliamHeader` to `SendCertificateChain`.
